File: agenkit-rust/examples/http_transport.rs
///! HTTP transport example.
///!
///! This example demonstrates how to:
///! 1. Create an agent server that exposes an agent over HTTP
///! 2. Create an HTTP client that communicates with the remote agent

use agenkit::core::{Agent, AgentError, Message};
use agenkit::transports::{HttpAgent, HttpServer, HttpTransportConfig};
use async_trait::async_trait;
use tokio::time::{sleep, Duration};

/// Counter agent that counts the number of messages it has processed.
struct CounterAgent {
    name: String,
    count: std::sync::Arc<std::sync::atomic::AtomicUsize>,
}

impl CounterAgent {
    fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            count: std::sync::Arc::new(std::sync::atomic::AtomicUsize::new(0)),
        }
    }
}

#[async_trait]
impl Agent for CounterAgent {
    fn name(&self) -> &str {
        &self.name
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        // Increment counter
        let count = self
            .count
            .fetch_add(1, std::sync::atomic::Ordering::SeqCst)
            + 1;

        // Return message with count
        let content = format!(
            "Message #{}: {}",
            count,
            message.content_as_str().unwrap_or("")
        );

        Ok(Message::with_text("assistant", content))
    }

    fn capabilities(&self) -> Vec<String> {
        vec!["counter".to_string()]
    }
}

#[tokio::main]
async fn main() {
    // Initialize tracing for debugging
    tracing_subscriber::fmt::init();

    // Create a counter agent
    let agent = CounterAgent::new("counter");

    // Start HTTP server in background
    let server = HttpServer::new(agent, "127.0.0.1:8080");
    tokio::spawn(async move {
        println!("Starting HTTP server on 127.0.0.1:8080");
        if let Err(e) = server.serve().await {
            eprintln!("Server error: {}", e);
        }
    });

    // Give server time to start
    sleep(Duration::from_millis(500)).await;

    // Create HTTP client
    let config = HttpTransportConfig {
        base_url: "http://127.0.0.1:8080".to_string(),
        timeout_secs: 30,
        api_key: None,
    };
    let client = HttpAgent::new("counter-client", config);

    // Send several messages
    for i in 1..=5 {
        let msg = Message::with_text("user", format!("Hello #{}", i));
        println!("\nSending: {}", msg.content_as_str().unwrap());

        match client.process(msg).await {
            Ok(response) => {
                println!("Response: {}", response.content_as_str().unwrap());
            }
            Err(e) => {
                eprintln!("Error: {}", e);
            }
        }

        sleep(Duration::from_millis(100)).await;
    }

    println!("\nHTTP transport example complete!");
}

File: agenkit-cpp/examples/other/http_transport.cpp
/**
 * @file http_transport.cpp
 * @brief Example demonstrating HTTP client/server communication
 *
 * This example shows how to:
 * - Start an HTTP server exposing an agent
 * - Create an HTTP client to communicate with the server
 * - Send messages and receive responses
 */

#include <iostream>
#include <thread>
#include <chrono>
#include "agenkit/adapters/echo_agent.hpp"
#include "agenkit/transports/http_server.hpp"
#include "agenkit/transports/http_agent.hpp"

using namespace agenkit;

void run_server() {
    std::cout << "[Server] Starting server..." << std::endl;

    // Create echo agent
    auto agent = std::make_shared<adapters::EchoAgent>();

    // Create HTTP server
    transports::HttpServer server(agent, "127.0.0.1:8080");

    std::cout << "[Server] Listening on http://127.0.0.1:8080" << std::endl;
    std::cout << "[Server] Endpoints:" << std::endl;
    std::cout << "[Server]   POST /process - Process messages" << std::endl;
    std::cout << "[Server]   GET  /health  - Health check" << std::endl;
    std::cout << std::endl;

    // Serve (blocking)
    server.serve();

    std::cout << "[Server] Stopped" << std::endl;
}

void run_client() {
    std::cout << "[Client] Waiting for server to start..." << std::endl;

    // Wait for server to start
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    std::cout << "[Client] Connecting to server..." << std::endl;

    // Create HTTP client
    transports::HttpTransportConfig config{
        "http://127.0.0.1:8080",
        30,
        std::nullopt
    };

    transports::HttpAgent client("remote-agent", config);

    std::cout << "[Client] Connected to: " << client.name() << std::endl;
    std::cout << "[Client] Capabilities: ";
    for (const auto& cap : client.capabilities()) {
        std::cout << cap << " ";
    }
    std::cout << std::endl << std::endl;

    // Send multiple messages
    std::vector<std::string> messages = {
        "Hello, server!",
        "How are you?",
        "This is message 3",
        "Testing HTTP transport",
        "Final message"
    };

    for (size_t i = 0; i < messages.size(); i++) {
        std::cout << "[Client] Sending message " << (i + 1) << ": \""
                  << messages[i] << "\"" << std::endl;

        auto msg = core::Message::with_text("user", messages[i]);
        msg.with_metadata("example", "http_transport")
           .with_metadata("message_id", static_cast<int>(i + 1));

        auto future = client.process(std::move(msg));
        auto result = future.get();

        if (result.is_ok()) {
            auto response = result.unwrap();
            std::cout << "[Client] Response: \"" << response.content_as_str() << "\""
                      << std::endl;
        } else {
            auto error = result.unwrap_err();
            std::cerr << "[Client] Error: " << error.message() << std::endl;
        }

        std::cout << std::endl;

        // Small delay between messages
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    std::cout << "[Client] All messages sent successfully!" << std::endl;
    std::cout << "[Client] Press Ctrl+C to stop server..." << std::endl;
}

int main() {
    std::cout << "=== Agenkit C++ HTTP Transport Example ===" << std::endl;
    std::cout << std::endl;

    // Start server in background thread
    std::thread server_thread(run_server);

    // Run client in main thread
    run_client();

    // In a real application, you'd have a proper shutdown mechanism
    // For this example, server will be terminated with Ctrl+C

    server_thread.detach();  // Detach server thread

    return 0;
}
